### PR TITLE
Update to Version 3.1

### PR DESCRIPTION
**Pull request by eggerk**
See https://github.com/ethz-asl/multiagent_mapping/pull/3302
Code adjusted from opencv2_catkin

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
-    -D BUILD_SHARED_LIBS=ON
-    -D BUILD_EXAMPLES=OFF
-    -D BUILD_PERF_TESTS=OFF
-    -D BUILD_TESTS=OFF
-    -D BUILD_FAT_JAVA_LIB=OFF
-    -D INSTALL_C_EXAMPLES=OFF
-    -D INSTALL_PYTHON_EXAMPLES=OFF
-    -D ENABLE_PRECOMPILED_HEADERS=OFF
-    -D CMAKE_INSTALL_PREFIX=${CATKIN_DEVEL_PREFIX} .
+    -DOPENCV_EXTRA_MODULES_PATH=../../../${CONTRIB_NAME}-prefix/src/${CONTRIB_NAME}/modules .
```
Comment (dymczykm) [inline CMakeLists.txt]: that's a bit scary, but I assume there's no other way of getting this path?

File: CMakeLists.txt [hunk 2]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
 ExternalProject_Add(opencv3_src
-  URL https://github.com/Itseez/opencv/archive/3.0.0-beta.zip
+  URL https://github.com/Itseez/opencv/archive/3.1.0.zip
   UPDATE_COMMAND ""
-  PATCH_COMMAND cd ../opencv3_src/
   CONFIGURE_COMMAND cd ../opencv3_src && cmake
-    -D WITH_QT=ON
-    -D WITH_OPENGL=ON
-    -D WITH_VTK=ON
-    -D WITH_OPENMP=ON
```
Comment (dymczykm) [inline CMakeLists.txt:20]: there's no more switch for openmp?

File: CMakeLists.txt [hunk 3]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
+    -DINSTALL_C_EXAMPLES=OFF
+    -DINSTALL_PYTHON_EXAMPLES=OFF
+    -DBUILD_EXAMPLES=OFF
+    -DWITH_QT=OFF # Needed by clang under Ubuntu 14.04 and GTK_WIDGET(cvGetWindowHandle(...)) with gcc (image_view)
+    -DWITH_OPENGL=ON
+    -DWITH_VTK=ON
+    -DENABLE_PRECOMPILED_HEADERS=OFF
+    -DCMAKE_INSTALL_PREFIX=${CATKIN_DEVEL_PREFIX}
+    -DBUILD_SHARED_LIBS=ON
+    -DWITH_CUDA=OFF
```
Comment (dymczykm) [inline CMakeLists.txt:42]: can this be turned on?

File: CMakeLists.txt [hunk 4]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
+    -DWITH_OPENCL=OFF
+    -DBUILD_opencv_gpu=OFF
```
Comment (dymczykm) [inline CMakeLists.txt:44]: can this be turned on?

File: CMakeLists.txt (hunk 3 above)
Comment (mfehr) [reply CMakeLists.txt:42]: wouldn't this create a dependency on cuda for everyone that links this package?

File: CMakeLists.txt [hunk 5]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
+    -DBUILD_opencv_ts=OFF
+    -DBUILD_opencv_hdf=OFF
+    -DBUILD_TESTS=OFF -DBUILD_opencv_ts=OFF .
   BUILD_COMMAND cd ../opencv3_src && make -j8
```
Comment (dymczykm) [inline CMakeLists.txt]: it might make sense to evaluate the number of cores our jenkins slaves have - some of them might have more than 8 physical cores

File: CMakeLists.txt [hunk 6]
```diff
@@ -9,38 +9,55 @@ include(ExternalProject)
...
 )
 
 cs_add_library(${PROJECT_NAME} src/dependency_tracker.cc)
 add_dependencies(${PROJECT_NAME} opencv3_src)
-target_link_libraries(${PROJECT_NAME} 
+target_link_libraries(${PROJECT_NAME}
   ${CATKIN_DEVEL_PREFIX}/lib/libopencv_core${CMAKE_SHARED_LIBRARY_SUFFIX})
 
 cs_install()
+install(DIRECTORY ${CATKIN_DEVEL_PREFIX}/include/opencv3
```
Comment (dymczykm) [inline CMakeLists.txt]: why is this necessary?

File: CMakeLists.txt (hunk 3 above)
Comment (dymczykm) [reply CMakeLists.txt:42]: if it does then OFF is fine, just asking :)

File: CMakeLists.txt (hunk 1 above)
Comment (eggerk) [reply CMakeLists.txt]: not sure, I can see if I can find a better way

File: CMakeLists.txt (hunk 2 above)
Comment (eggerk) [reply CMakeLists.txt:20]: I actually copied the whole project from `opencv2_catkin` and changed the url and added the extra contrib package.
(same for all comments below)

File: CMakeLists.txt (hunk 4 above)
Comment (eggerk) [reply CMakeLists.txt:44]: I can still compile with this (and cuda) turned on. But I didn't test/compile all the way to swe_app (working on muma2 now). I also don't know what happens when you try to compile it on a computer without gpu. It's also disabled in opencv2_catkin

File: CMakeLists.txt (hunk 6 above)
Comment (eggerk) [reply CMakeLists.txt]: Seems to compile on my computer without this. I just copied it from opencv2_catkin

File: CMakeLists.txt (hunk 3 above)
Comment (eggerk) [reply CMakeLists.txt:42]: Appearantly, the cuda module is built even if it's set to off: http://docs.opencv.org/3.1.0/d2/dbc/cuda_intro.html
This would just enable the use of cuda functionality. So I guess we could turn this to on if it's necessary.
@dymczykm wdyt?

File: CMakeLists.txt (hunk 4 above)
Comment (dymczykm) [reply CMakeLists.txt:44]: ok, let's leave it as is

File: CMakeLists.txt (hunk 3 above)
Comment (dymczykm) [reply CMakeLists.txt:42]: ok, let's leave it as is

**Review by dymczykm — APPROVED** [review]
lgtm
